### PR TITLE
Display substack failures when pcluster create fails

### PR DESCRIPTION
**Pull request by rexcsn**
* Add logic to display detailed substack failure message when pcluster creation fails due to a substack error.
* Example of old message:
```
Creating stack named: parallelcluster-fsx-fail-old
Status: parallelcluster-fsx-fail-old - CREATE_FAILED                            
Cluster creation failed.  Failed events:
  - AWS::CloudFormation::Stack parallelcluster-fsx-fail-old The following resource(s) failed to create: [FSXSubstack, AssociateEIP, RootInstanceProfile, EBSCfnStack]. 
  - AWS::CloudFormation::Stack EBSCfnStack Resource creation cancelled
  - AWS::IAM::InstanceProfile RootInstanceProfile Resource creation cancelled
  - AWS::EC2::EIPAssociation AssociateEIP Resource creation cancelled
  - AWS::CloudFormation::Stack FSXSubstack Embedded stack arn:aws:cloudformation:us-east-2:704743599507:stack/parallelcluster-fsx-fail-old-FSXSubstack-1SJCNA5YGN9Z4/40e7aee0-5278-11ea-a633-064ea827a74e was not successfully created: The following resource(s) failed to create: [FileSystem]. 
```
* Example of new message:
```
Creating stack named: parallelcluster-fsx-fail
Status: parallelcluster-fsx-fail - CREATE_FAILED                                
Cluster creation failed.  Failed events:
  - AWS::CloudFormation::Stack parallelcluster-fsx-fail The following resource(s) failed to create: [FSXSubstack, AssociateEIP, RootInstanceProfile, EBSCfnStack]. 
  - AWS::CloudFormation::Stack EBSCfnStack Resource creation cancelled
  - AWS::EC2::EIPAssociation AssociateEIP Resource creation cancelled
  - AWS::IAM::InstanceProfile RootInstanceProfile Resource creation cancelled
  - AWS::CloudFormation::Stack FSXSubstack Embedded stack arn:aws:cloudformation:us-east-2:704743599507:stack/parallelcluster-fsx-fail-FSXSubstack-PJZKIVCKR2WQ/d0f95b10-5277-11ea-902e-023006f8d510 was not successfully created: The following resource(s) failed to create: [FileSystem]. 
    - AWS::CloudFormation::Stack parallelcluster-fsx-fail-FSXSubstack-PJZKIVCKR2WQ The following resource(s) failed to create: [FileSystem]. 
    - AWS::FSx::FileSystem FileSystem Invalid storage capacity provided: 0. Minimum storage capacity for LUSTRE file systems is 1200 in this region. (Service: AmazonFSx; Status Code: 400; Error Code: BadRequest; Request ID: 0137e76b-5eef-4720-a9a4-f719814160d7)
```

Signed-off-by: Rex <shuningc@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/pcluster/utils.py [hunk 1]
```diff
@@ -393,19 +394,41 @@ def verify_stack_creation(stack_name, cfn_client):
...
+        if event.get("ResourceStatus") == "CREATE_FAILED":
+            _log_failed_cfn_event(event, indent=indent)
+            if event.get("ResourceType") == "AWS::CloudFormation::Stack":
+                # Sample substack error:
+                # "Embedded stack arn:aws:cloudformation:us-east-2:704743599507:stack/
+                # parallelcluster-fsx-fail-FSXSubstack-65ITLJEZJ0DQ/
+                # 3a4ecf00-51e7-11ea-8e3e-022fd555c652 was not successfully created:
+                # The following resource(s) failed to create: [FileSystem]."
+                substack_capture = re.search(
+                    ".+/(parallelcluster-.+)/".format(stack_name), event.get("ResourceStatusReason")
```
Comment (enrico-usai) [inline cli/pcluster/utils.py]: You're using `format` but there are no placeholders to replace in the string.
Then we should use the `PCLUSTER_STACK_PREFIX` variable from the constants.py.

File: cli/pcluster/utils.py [hunk 2]
```diff
@@ -393,19 +394,41 @@ def verify_stack_creation(stack_name, cfn_client):
...
+                )
+                substack_name = substack_capture.group(1) if substack_capture else None
+                if substack_name:
+                    _log_stack_failure_recursive(substack_name, indent=indent + 2)
+
+
+def _log_failed_cfn_event(event, indent):
```
Comment (enrico-usai) [inline cli/pcluster/utils.py]: let's put 2 as default here, in this way you can remove the parameter in both rows 397 and 407

File: cli/pcluster/utils.py [hunk 3]
```diff
@@ -393,19 +394,41 @@ def verify_stack_creation(stack_name, cfn_client):
...
-        events = get_stack_events(stack_name, raise_on_error=True)
-        for event in events:
-            if event.get("ResourceStatus") == "CREATE_FAILED":
-                LOGGER.info(
-                    "  - %s %s %s",
-                    event.get("ResourceType"),
-                    event.get("LogicalResourceId"),
-                    event.get("ResourceStatusReason"),
-                )
+        _log_stack_failure_recursive(stack_name, indent=2)
```
Comment (enrico-usai) [inline cli/pcluster/utils.py]: Could you add an example of output in the PR description before and after your patch?

File: cli/pcluster/utils.py [hunk 4]
```diff
@@ -393,19 +394,41 @@ def verify_stack_creation(stack_name, cfn_client):
...
+    for event in events:
+        if event.get("ResourceStatus") == "CREATE_FAILED":
+            _log_failed_cfn_event(event, indent=indent)
+            if event.get("ResourceType") == "AWS::CloudFormation::Stack":
+                # Sample substack error:
+                # "Embedded stack arn:aws:cloudformation:us-east-2:704743599507:stack/
+                # parallelcluster-fsx-fail-FSXSubstack-65ITLJEZJ0DQ/
+                # 3a4ecf00-51e7-11ea-8e3e-022fd555c652 was not successfully created:
+                # The following resource(s) failed to create: [FileSystem]."
+                substack_capture = re.search(
```
Comment (enrico-usai) [inline cli/pcluster/utils.py]: substack_capture --> substack_error ?

File: cli/pcluster/utils.py [hunk 5]
```diff
@@ -393,19 +394,41 @@ def verify_stack_creation(stack_name, cfn_client):
...
+                )
+                substack_name = substack_capture.group(1) if substack_capture else None
+                if substack_name:
+                    _log_stack_failure_recursive(substack_name, indent=indent + 2)
+
+
+def _log_failed_cfn_event(event, indent):
+    """Log failed CFN events."""
+    LOGGER.info(
+        (" " * indent + "- %s %s %s"),
```
Comment (enrico-usai) [inline cli/pcluster/utils.py]: let's use format here too.

File: cli/pcluster/utils.py [hunk 6]
```diff
@@ -393,19 +394,41 @@ def verify_stack_creation(stack_name, cfn_client):
...
+                substack_name = substack_error.group(1) if substack_error else None
+                if substack_name:
+                    _log_stack_failure_recursive(substack_name, indent=indent + 2)
+
+
+def _log_failed_cfn_event(event, indent):
+    """Log failed CFN events."""
+    LOGGER.info(
+        " " * indent
+        + "- {0} {1} {2}".format(
```
Comment (enrico-usai) [inline cli/pcluster/utils.py]: I'd write:
```python
"{0}- {1} {2} {3}".format(" " * indent, ...)
```

Comment (rexcsn) [reply cli/pcluster/utils.py]: I think we should use string interpolation to conform to pylint: logging-format-interpolation (W1202).
```
LOGGER.info(
        "%s- %s %s %s",
        " " * indent,
        event.get("ResourceType"),
        event.get("LogicalResourceId"),
        event.get("ResourceStatusReason"),
)
```

**Pull request merged**